File: timesheet-example/workflows-java/src/test/java/com/example/test/flow/IssueInvoiceFlowTests.java
package com.example.test.flow;


import com.google.common.collect.ImmutableList;
import com.example.flow.IssueInvoiceFlow;
import com.example.flow.QueryRateHandler;
import com.example.state.InvoiceState;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.TransactionState;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.transactions.SignedTransaction;
import net.corda.testing.node.MockNetwork;
import net.corda.testing.node.MockNetworkParameters;
import net.corda.testing.node.StartedMockNode;
import net.corda.testing.node.TestCordapp;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.security.SignatureException;
import java.time.LocalDate;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class IssueInvoiceFlowTests {
    private MockNetwork network;
    private StartedMockNode contractor;
    private StartedMockNode megaCorp;
    private StartedMockNode oracle;
    private LocalDate today;

    @Before
    public void setup() {
        network = new MockNetwork(new MockNetworkParameters(
                ImmutableList.of(
                        TestCordapp.findCordapp("com.example.contract"),
                        TestCordapp.findCordapp("com.example.flow"),
                        TestCordapp.findCordapp("com.example.state"),
                        TestCordapp.findCordapp("com.example.service")
                )
        ));
        CordaX500Name oracleName = new CordaX500Name("Oracle", "London", "GB");
        contractor = network.createPartyNode(null);
        megaCorp = network.createPartyNode(null);
        oracle = network.createPartyNode(oracleName);
        // For real nodes this happens automatically, but we have to manually register the flow for tests.
        ImmutableList.of(contractor, megaCorp).forEach(it -> {
            it.registerInitiatedFlow(IssueInvoiceFlow.Acceptor.class);
            it.registerInitiatedFlow(QueryRateHandler.class);
        });
        network.runNetwork();
        today = LocalDate.now();
    }

    @After
    public void tearDown(){
        network.stopNodes();
    }

    @Test(expected = ExecutionException.class)
    public void flowRejectsInvalidInvoices() throws ExecutionException, InterruptedException {
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(-1, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);

        network.runNetwork();
        future.get();
    }

    @Test
    public void SignedTransactionsReturnedByFlowIsSignedByInitiator() throws ExecutionException, InterruptedException, SignatureException {
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(1, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);
        network.runNetwork();

        SignedTransaction stx = (SignedTransaction) future.get();
        stx.verifySignaturesExcept(megaCorp.getInfo().getLegalIdentities().get(0).getOwningKey());
    }

    @Test
    public void SignedTRansactionReturnedByFlowIsSignedByAcceptor() throws SignatureException, ExecutionException, InterruptedException {
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(1, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);
        network.runNetwork();

        SignedTransaction stx = (SignedTransaction) future.get();
        stx.verifySignaturesExcept(contractor.getInfo().getLegalIdentities().get(0).getOwningKey());
    }

    @Test
    public void FlowRecordsATransationInBothPartiesTransactioStorage() throws ExecutionException, InterruptedException {
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(1, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);
        network.runNetwork();

        SignedTransaction stx = (SignedTransaction) future.get();

        for (StartedMockNode node: ImmutableList.of(contractor, megaCorp)
             ) {
            Assert.assertEquals(stx, node.getServices().getValidatedTransactions().getTransaction(stx.getId()));

        }
    }

    @Test
    public void recordedTransactionHAsNoIputsAndASingleOutput() throws ExecutionException, InterruptedException {
        int invoiceValue = 1;
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(invoiceValue, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);
        network.runNetwork();

        SignedTransaction stx = (SignedTransaction) future.get();

        for (StartedMockNode node: ImmutableList.of(contractor, megaCorp)
        ) {
            SignedTransaction rtx = node.getServices().getValidatedTransactions().getTransaction(stx.getId());
            List<TransactionState<ContractState>> txOutputs = rtx.getTx().getOutputs();
            assert(txOutputs.size() == 1);

            InvoiceState recordedState = (InvoiceState) txOutputs.get(0).getData();

            Assert.assertEquals(recordedState.getHoursWorked(), invoiceValue);
            Assert.assertEquals(recordedState.getContractor(), contractor.getInfo().getLegalIdentities().get(0));
            Assert.assertEquals(recordedState.getCompany(), megaCorp.getInfo().getLegalIdentities().get(0));
        }
    }

    @Test
    public void flowRecordsTheCorrectInvoicesInBothPartiesVaults() throws ExecutionException, InterruptedException {
        int invoiceValue = 1;
        IssueInvoiceFlow.Initiator flow = new IssueInvoiceFlow.Initiator(invoiceValue, today, megaCorp.getInfo().getLegalIdentities().get(0));
        Future future = contractor.startFlow(flow);
        network.runNetwork();

        SignedTransaction stx = (SignedTransaction) future.get();

        for (StartedMockNode node: ImmutableList.of(contractor, megaCorp)
        ) {
            node.transaction(() -> {
                   List<StateAndRef<InvoiceState>>  invoices = node.getServices().getVaultService().queryBy(InvoiceState.class).getStates();
                   Assert.assertEquals(1, invoices.size());
                   InvoiceState recordedState = invoices.get(0).getState().getData();
                   Assert.assertEquals(recordedState.getHoursWorked(), invoiceValue);
                   Assert.assertEquals(recordedState.getDate(), today);
                   Assert.assertEquals(recordedState.getContractor(), contractor.getInfo().getLegalIdentities().get(0));
                   Assert.assertEquals(recordedState.getCompany(), megaCorp.getInfo().getLegalIdentities().get(0));
                   return null;
            });

        }
    }
}

File: timesheet-example/workflows-kotlin/src/test/kotlin/com/example/test/flow/IssueInvoiceFlowTests.kt
package com.example.test.flow

import com.example.flow.IssueInvoiceFlow
import com.example.state.InvoiceState
import net.corda.core.contracts.TransactionVerificationException
import net.corda.core.identity.CordaX500Name
import net.corda.core.node.services.queryBy
import net.corda.core.utilities.getOrThrow
import net.corda.testing.core.singleIdentity
import net.corda.testing.node.MockNetwork
import net.corda.testing.node.MockNetworkParameters
import net.corda.testing.node.StartedMockNode
import net.corda.testing.node.TestCordapp
import org.junit.After
import org.junit.Before
import org.junit.Test
import java.time.LocalDate
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class IssueInvoiceFlowTests {
    private lateinit var network: MockNetwork
    private lateinit var contractor: StartedMockNode
    private lateinit var megaCorp: StartedMockNode
    private lateinit var oracle: StartedMockNode
    private lateinit var today: LocalDate

    @Before
    fun setup() {
        network = MockNetwork(MockNetworkParameters(cordappsForAllNodes = listOf(
                TestCordapp.findCordapp("com.example.contract"),
                TestCordapp.findCordapp("com.example.flow")
        )))
        val oracleName = CordaX500Name("Oracle", "London","GB")
        contractor = network.createPartyNode()
        megaCorp = network.createPartyNode()
        oracle = network.createNode(oracleName)
        // For real nodes this happens automatically, but we have to manually register the flow for tests.
        listOf(contractor, megaCorp).forEach { it.registerInitiatedFlow(IssueInvoiceFlow.Acceptor::class.java) }
        network.runNetwork()
        today = LocalDate.now()
    }

    @After
    fun tearDown() {
        network.stopNodes()
    }

    @Test
    fun `flow rejects invalid invoices`() {
        val flow = IssueInvoiceFlow.Initiator(-1, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()

        // The InvoiceContract specifies that hoursWorked cannot have negative values.
        assertFailsWith<TransactionVerificationException> { future.getOrThrow() }
    }

    @Test
    fun `SignedTransaction returned by the flow is signed by the initiator`() {
        val flow = IssueInvoiceFlow.Initiator(1, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()

        val signedTx = future.getOrThrow()
        signedTx.verifySignaturesExcept(megaCorp.info.singleIdentity().owningKey)
    }

    @Test
    fun `SignedTransaction returned by the flow is signed by the acceptor`() {
        val flow = IssueInvoiceFlow.Initiator(1, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()

        val signedTx = future.getOrThrow()
        signedTx.verifySignaturesExcept(contractor.info.singleIdentity().owningKey)
    }

    @Test
    fun `flow records a transaction in both parties' transaction storage`() {
        val flow = IssueInvoiceFlow.Initiator(1, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()
        val signedTx = future.getOrThrow()

        // We check the recorded transaction in both transaction storages.
        for (node in listOf(contractor, megaCorp)) {
            assertEquals(signedTx, node.services.validatedTransactions.getTransaction(signedTx.id))
        }
    }

    @Test
    fun `recorded transaction has no inputs and a single output, the input invoice`() {
        val invoiceValue = 1
        val flow = IssueInvoiceFlow.Initiator(invoiceValue, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()
        val signedTx = future.getOrThrow()

        // We check the recorded transaction in both vaults.
        for (node in listOf(contractor, megaCorp)) {
            val recordedTx = node.services.validatedTransactions.getTransaction(signedTx.id)
            val txOutputs = recordedTx!!.tx.outputs
            assert(txOutputs.size == 1)

            val recordedState = txOutputs[0].data as InvoiceState
            assertEquals(recordedState.hoursWorked, invoiceValue)
            assertEquals(recordedState.contractor, contractor.info.singleIdentity())
            assertEquals(recordedState.company, megaCorp.info.singleIdentity())
        }
    }

    @Test
    fun `flow records the correct invoice in both parties' vaults`() {
        val invoiceValue = 1
        val flow = IssueInvoiceFlow.Initiator(1, today, megaCorp.info.singleIdentity())
        val future = contractor.startFlow(flow)
        network.runNetwork()
        future.getOrThrow()

        // We check the recorded invoice in both vaults.
        for (node in listOf(contractor, megaCorp)) {
            node.transaction {
                val invoices = node.services.vaultService.queryBy<InvoiceState>().states
                assertEquals(1, invoices.size)
                val recordedState = invoices.single().state.data
                assertEquals(recordedState.hoursWorked, invoiceValue)
                assertEquals(recordedState.date, today)
                assertEquals(recordedState.contractor, contractor.info.singleIdentity())
                assertEquals(recordedState.company, megaCorp.info.singleIdentity())
            }
        }
    }
}